File: app/packages/magicblock_client/lib/src/oracle.dart
/// Parses a MagicBlock ephemeral-oracle `PriceUpdateV2` account (Pyth Lazer feeds, in the ER) by
/// raw byte offset — the SAME layout the on-chain `throtl-engine` reads (oracle.rs). Keep them in
/// lock-step: price i64@73, conf u64@81, expo i32@89, publish_time i64@93 (after the 8-byte disc).
library;

import 'dart:typed_data';

import 'package:meta/meta.dart';

/// Oracle program id (devnet + mainnet) and the live devnet SOL feed PDA (verified actively ticking).
const String oracleProgramId = 'PriCems5tHihc6UDXDjzjeawomAwBduWMGAi8ZUjppd';
const String devnetSolPriceFeedPda = 'ENYwebBThHzmzwPLAQvCucUTsjyfBSZdD9ViXksS4jPu';

/// The feed-id seed encoding: the Pyth Lazer id rendered as a DECIMAL STRING (SOL lazerId = 6 → "6").
String lazerFeedSeed(int lazerId) => '$lazerId';

const _offPrice = 73;
const _offExpo = 89;
const _offPublishTime = 93;
const int _minLen = _offPublishTime + 8; // 101

@immutable
class OraclePrice {
  const OraclePrice({required this.priceE9, required this.expo, required this.publishTime});

  /// Price normalized to 1e9 fixed-point (matches the program's `price_e9`).
  final int priceE9;
  final int expo;
  final int publishTime;

  double get ui => priceE9 / 1000000000;
}

/// Decode a `PriceUpdateV2` account's data bytes. Returns null if too short / non-positive.
OraclePrice? parsePriceUpdateV2(Uint8List data) {
  if (data.length < _minLen) return null;
  final bd = ByteData.sublistView(data);
  final priceRaw = bd.getInt64(_offPrice, Endian.little);
  final expo = bd.getInt32(_offExpo, Endian.little);
  final publishTime = bd.getInt64(_offPublishTime, Endian.little);
  if (priceRaw <= 0) return null;
  final priceE9 = _normalizeToE9(priceRaw, expo);
  if (priceE9 == null || priceE9 <= 0) return null;
  return OraclePrice(priceE9: priceE9, expo: expo, publishTime: publishTime);
}

/// `price * 10^(9 - expo)` → int (price_e9).
///
/// IMPORTANT: MagicBlock's ephemeral-oracle (PriceUpdateV3) stores the **Pyth Lazer** exponent,
/// which is a POSITIVE number of decimals: `real = mantissa × 10^(-exponent)`. Verified live on the
/// devnet SOL feed: mantissa 6786596052, exponent 8 → $67.866. (NOT the standard Pyth `× 10^+expo`.)
/// Uses BigInt to match the program's wide intermediate exactly.
int? _normalizeToE9(int price, int expo) {
  final target = 9 - expo;
  final p = BigInt.from(price);
  final BigInt v;
  if (target >= 0) {
    v = p * BigInt.from(10).pow(target);
  } else {
    v = p ~/ BigInt.from(10).pow(-target);
  }
  if (v > (BigInt.one << 63) - BigInt.one) return null;
  return v.toInt();
}

File: programs/throtl-engine/programs/throtl-engine/src/oracle.rs
//! Reads a MagicBlock ephemeral-oracle `PriceUpdateV3` PDA (Pyth Lazer feeds, in the ER) by raw
//! byte offset — no `pyth-solana-receiver-sdk` / Anchor dependency on the consumer side.
//!
//! Layout (Borsh, after the 8-byte Anchor account discriminator; verified live on the devnet SOL
//! feed `ENYweb…`): `8 disc | 32 write_authority | 1 verification_level | 32 feed_id |
//! i64 price@73 | u64 conf@81 | i32 exponent@89 | i64 publish_time@93 | …`.
//!
//! EXPONENT CONVENTION (the trap that the M0.1 spike caught): this oracle stores the **Pyth Lazer**
//! exponent — a POSITIVE number of decimals where `real = mantissa × 10^(-exponent)` — NOT the
//! standard Pyth signed exponent. Live SOL: mantissa 6_786_596_052, exponent 8 → $67.866. So we
//! normalize with `10^(9 - exponent)`, not `10^(9 + exponent)`.

use anchor_lang::prelude::*;

use crate::constants::{MAX_ORACLE_STALENESS_SECS, ORACLE_PROGRAM_ID};
use crate::error::ThrotlError;

const OFF_FEED_ID: usize = 8 + 32 + 1; // 41
const OFF_PRICE: usize = 73;
const OFF_EXPO: usize = 89;
const OFF_PUBLISH_TIME: usize = 93;
const MIN_LEN: usize = OFF_PUBLISH_TIME + 8; // 101

pub struct OraclePrice {
    /// Price normalized to 1e9 fixed-point.
    pub price_e9: u64,
    pub publish_time: i64,
}

/// Validate ownership + feed + staleness, then return the mark normalized to price_e9.
pub fn read_oracle(
    acc: &AccountInfo,
    expected_feed_id: &[u8; 32],
    now: i64,
) -> Result<OraclePrice> {
    require_keys_eq!(
        *acc.owner,
        ORACLE_PROGRAM_ID,
        ThrotlError::OracleOwnerMismatch
    );

    let data = acc.try_borrow_data()?;
    require!(data.len() >= MIN_LEN, ThrotlError::OracleInvalidPrice);

    require!(
        data[OFF_FEED_ID..OFF_FEED_ID + 32] == expected_feed_id[..],
        ThrotlError::OracleFeedMismatch
    );

    let price_raw = i64::from_le_bytes(data[OFF_PRICE..OFF_PRICE + 8].try_into().unwrap());
    let expo = i32::from_le_bytes(data[OFF_EXPO..OFF_EXPO + 4].try_into().unwrap());
    let publish_time = i64::from_le_bytes(
        data[OFF_PUBLISH_TIME..OFF_PUBLISH_TIME + 8]
            .try_into()
            .unwrap(),
    );

    require!(price_raw > 0, ThrotlError::OracleInvalidPrice);
    require!(
        now - publish_time <= MAX_ORACLE_STALENESS_SECS,
        ThrotlError::OracleStale
    );

    let price_e9 = normalize_to_e9(price_raw as u128, expo).ok_or(ThrotlError::MathOverflow)?;
    require!(price_e9 > 0, ThrotlError::OracleInvalidPrice);

    Ok(OraclePrice {
        price_e9,
        publish_time,
    })
}

/// `price * 10^(9 - expo)` → u64 (price_e9). `expo` is the POSITIVE Lazer exponent (decimals);
/// `real = mantissa × 10^(-expo)`. See the module header.
fn normalize_to_e9(price: u128, expo: i32) -> Option<u64> {
    let target = 9i32 - expo;
    let v = if target >= 0 {
        price.checked_mul(10u128.checked_pow(target as u32)?)?
    } else {
        price.checked_div(10u128.checked_pow((-target) as u32)?)?
    };
    if v > u64::MAX as u128 {
        None
    } else {
        Some(v as u64)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_lazer_expo_8() {
        // Live devnet SOL: mantissa 6_786_596_052, Lazer exponent 8 → price_e9 = raw * 10.
        assert_eq!(normalize_to_e9(6_786_596_052, 8), Some(67_865_960_520));
    }

    #[test]
    fn normalize_expo_9_identity() {
        assert_eq!(normalize_to_e9(147_123_456_789, 9), Some(147_123_456_789));
    }

    #[test]
    fn normalize_expo_10_divides() {
        assert_eq!(
            normalize_to_e9(1_471_234_567_890, 10),
            Some(147_123_456_789)
        );
    }
}
